### PR TITLE
doc: Explain squashing with merge commits

### PR DESCRIPTION
**Pull request by maflcko**
This avoids having to explain it in each thread

### CODE REVIEW

**Comment by MarcoFalke** [comment]
:rotating_light: Warning, this is untested, but tests and review are welcome.

**Review by InvestorMoney — APPROVED** [review]

**Review by david-bakin — COMMENTED** [review]
What's wrong with a `git merge --squash` which I see is _not_ recommended earlier in this document or in your update?  (I really don't know, BTW, which is why I'm asking.)

**Review by jonatack — COMMENTED** [review]
-0 on adding further git documentation to the repo, there are other and better resources

**Comment by jonatack** [comment]
> This avoids having to explain it in each thread

AFAICS you use pre-prepared maintainer responses and could add an appropriate link to an external resource where needed.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -216,9 +216,25 @@ Please update the resulting commit message, if needed. It should read as a
...
 commits.
 
-If you have problems with squashing or other git workflows, you can enable
-"Allow edits from maintainers" in the right-hand sidebar of the GitHub web
-interface and ask for help in the pull request.
+If your change contain a merge commit, the above workflow may not work. You
+have the option to squash everything into *one* commit, or *keep* the commits
+in a non-interactive rebase.
+
+To squash everything into one commit.
```
Comment (laanwj) [inline CONTRIBUTING.md:223]: I think keeping this part is enough. By far and wide, a new contributor will want to squash everything into one commit.

**Comment by laanwj** [comment]
Concept ACK. I think having some easy to follow steps here is useful. I do agree with @jonatack that going into too much details about git usage is out of scope.

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/25165/commits/faa280a9d8084786e1fd5f82b3b6a5a45652ade7

LGTM

**Comment by MarcoFalke** [comment]
> What's wrong with a git merge --squash which I see is not recommended earlier in this document or in your update? (I really don't know, BTW, which is why I'm asking.)

For me, this does not work:

```
$ git merge --squash bitcoin-core/master 
Already up to date. (nothing to squash)

**Comment by MarcoFalke** [comment]
> -0 on adding further git documentation to the repo, there are other and better resources


I agree. Mind sharing a link to the "other and better resources" that can be used?

**Comment by MarcoFalke** [comment]
Recall that we are probably the only open source project that doesn't use the "squash on merge" GitHub feature, so I don't think there is documentation about this out there, but I am more than happy to be proven wrong.

File: CONTRIBUTING.md (hunk 1 above)
Comment (MarcoFalke) [reply CONTRIBUTING.md:220]: Thx, fixed.

**Review by jonatack — COMMENTED** [review]
> I agree. Mind sharing a link to the "other and better resources" that can be used?

A prepared response may be a better place than the repo for an external link, but here are a couple that are similar to your method (rather than, say, the more flexible `git rebase -i`).

- https://stackoverflow.com/a/32607742
- https://stackoverflow.com/a/5201642

**Comment by suhailsaqan** [comment]
> Changed to a more general approach (that may involve fixing conflicts again)

That didn't work for my case in #25161 unfortunately.

**Comment by MarcoFalke** [comment]
Can you elaborate a bit on this? I tried it on https://github.com/bitcoin/bitcoin/pull/25161 and it worked for me.

**Comment by suhailsaqan** [comment]
I must have done something wrong then. I can't remember exactly what happened but I think I actually accidentally pulled and pushed after running `git rebase FETCH_HEAD` which messed it all up again so I had to do something longer to fix it. Sorry for the confusion.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -216,9 +216,11 @@ Please update the resulting commit message, if needed. It should read as a
...
 commits.
 
-If you have problems with squashing or other git workflows, you can enable
-"Allow edits from maintainers" in the right-hand sidebar of the GitHub web
-interface and ask for help in the pull request.
+If your change contains a merge commit, the above workflow may not work and you
+will need to remove the merge commit first.
+
+    git fetch https://github.com/bitcoin/bitcoin  # Fetch the latest upstream commit
+    git rebase FETCH_HEAD
```
Comment (laanwj) [inline CONTRIBUTING.md:223]: Does a plain `rebase` work to remove merge commits? I always assumed it was much more hassle.

Comment (MarcoFalke) [reply CONTRIBUTING.md:223]: Yeah, it should remove them. Obviously you'll have to solve the conflicts again (if there were any solved in the merge commit).

Though I think this general approach should work in all possible cases regardless of how messed up  the commit history is. Any other approach requires a bit more git knowledge, but there were complaints that we don't want to dive too deep in this document.

**Comment by theStack** [comment]
> If your change contains a merge commit...

Very likely I'm missing something fundamental here, but why would a change in a PR ever contain a merge commit? Is this for the scenario of "PR foo is based on PR bar, PR bar has been merged, now PR foo needs to rebase?"

**Comment by MarcoFalke** [comment]
It usually happens when the author merges with current Bitcoin Core `master` instead of rebase.

**Review by theStack — APPROVED** [review]
ACK 888890302edacf3392957ef9e682cd6cb17a5f5c

Alternatively, could move the two git lines to the rebase section below (where currently concrete git instructions are missing, it only says that "git rebase" can be used, but without example) and link to there, e.g. "you will need to remove the merge commit first by rebasing on the target branch".

File: CONTRIBUTING.md (hunk 2 above)
Comment (MarcoFalke) [reply CONTRIBUTING.md:223]: Resolving as fixed?

**Comment by MarcoFalke** [comment]
Thx, done.

**Comment by jonatack** [comment]
Aside, just came across https://github.com/lsilva01/operating-bitcoin-core-v1/blob/main/git-tutorial.md ... maybe reply to people with https://github.com/lsilva01/operating-bitcoin-core-v1/blob/main/git-tutorial.md#2-squashing-and-rebasing when useful. (I'm not sure how sustainable it is for a contributor to not become familiar with `git rebase -i`, and most of the better tutorials and blog posts when searching for "how to squash your commits" seem to cover that approach, so suggesting it may be helpful to people, I don't know.)

**Comment by MarcoFalke** [comment]
Yeah, I am happy to review a pull request that removes the sections in our readme and replaces them with a link. Or should I do it here?

**Comment by laanwj** [comment]
ACK fa2d226ac950d8b4f7e430732f13ad408c504745
IMO, this is an improvement. The high-level longer term discussion on the scope of our documentation can of course go on.

**Pull request merged**